### ISSUE TITLE
Refactoring AD Tests

### ISSUE THREAD

**Issue by willtebbutt** [TuringLang/Turing.jl#2307]
Currently, there are many situation in which we run an entire sampler in order to test AD, such as many cases in https://github.com/TuringLang/Turing.jl/blob/master/test/mcmc/abstractmcmc.jl .

This is quite compute intensive. @yebai pointed out that perhaps we should just check that the gradient runs and is correct, rather than thousants of mcmc chains.

**Comment by penelopeysm** [comment]
Thinking about this and #2338 a little, it's not very easy to dive into the internals of `sample()` to check that the gradient is correct, but what I'm thinking we could do is:

1. seed (with StableRNGs) and sample for something like 10 iterations per AD backend and check that the results are numerically similar
2. once we've established that all AD backends give the same result, we choose one and run the full 1000 samples to check that the result is correct

I don't know how much this will really speed it up, but it could be worth a shot. Thoughts ... let me use the magic ping @TuringLang/maintainers?

**Comment by willtebbutt** [comment]
I agree with the general sentiment here. In my view we should
1. only test long-run sampling (i.e. to test the correctness of a sampler) once, using whichever AD we prefer.
2. publish an iterable collection of test cases that Turing.jl declares that it wants to be able to differentiate. This would form a single source of truth for "things Turing.jl would like to be able to differentiate" (which can be updated over time as we find problem cases / find new cases that we care about). This way, AD packages have clear guidance as to the things that Turing.jl wants to be able to differentiate, and can just hook in them in their integration tests. For example, @mhauru recently added a collection of test cases for Bijectors [here](https://github.com/compintell/Mooncake.jl/blob/main/test/integration_testing/bijectors/bijectors.jl) in Mooncake -- the key point here is that there's a single vector of test cases (see the `test_cases` variable) which Mooncake.jl can just iterate through.

It might be that we want to run these test cases in Turing.jl for each backend that we claim supports Turing.jl, in addition to running the tests in Mooncake.jl, but that's something that we could run later on.

Either way, I really think we should decouple "is this sampler correct?" from "does this AD work correctly?".

**Comment by penelopeysm** [comment]
I'll start playing around with this in a new package, we can decide whether to put it in DPPL proper at a later point in time.

**Comment by willtebbutt** [comment]
Agreed. We should think about the precise format though. For the sake of both running the tests in downstream AD packages, and for running them in the Turing.jl integration tests, might it make sense to produce things which can be consumed by DifferentiationInterface's testing functionality? See https://juliadiff.org/DifferentiationInterface.jl/DifferentiationInterfaceTest/dev/ for context.

**Comment by penelopeysm** [comment]
Ooh. I like that. But I don't think LogDensityProblemsAD uses DI for all its backends just yet, right? In which case, I _think_ we'd still need to test `logdensity_and_gradient`? (Please correct if wrong.) What we can do is to try to future proof it (if we assume that eventually Turing will be using DI for everything, e.g. if we define our own logdensitymodel interface) by making sure we don't stray too far from the DITest interface.

**Comment by willtebbutt** [comment]
Hmmm maybe you're right? That being said, LogDensityProblemsAD does now support arbitrary `ADTypes`, meaning that you can use `ADTypes` to make any backend supported by DifferentiationInterface work for you. e.g.
```julia
using ADTypes, Mooncake, Turing
NUTS(; adtype=ADTypes.AutoMooncake(; config=nothing))
```
so maybe it's fine?

**Comment by penelopeysm** [comment]
Mostly notes for myself because Julia really lends itself to creating a real maze of function calls, but if anyone's curious. 

Mooncake goes via DI so that's all fine.

However, passing an adtype of e.g. AutoForwardDiff still uses the custom implementation, instead of going via DifferentiationInterface:

Running `ADgradient(AutoForwardDiff(), ldf)` where `ldf isa DynamicPPL.LogDensityFunction`
 - hits this https://github.com/TuringLang/DynamicPPL.jl/blob/f5890a11ef4002026e8edeb355b6e8a50c68d115/ext/DynamicPPLForwardDiffExt.jl#L16-L36, which calls `ADgradient(:ForwardDiff, ...)`
 - which will then call https://github.com/tpapp/LogDensityProblemsAD.jl/blob/e3401f21b5a065df0d5de38b37fad0e6650618f3/ext/LogDensityProblemsADForwardDiffExt.jl#L96-L118.

It's possible to force the use of DI for AutoForwardDiff by:
 - modifying the DPPL code above to call `ADgradient(AutoForwardDiff(, ...), ...)`
 - that alone would still fail though, because LogDensityProblemsAD special-cases ForwardDiff, so even if you pass it AutoForwardDiff it won't use DI due to this function https://github.com/tpapp/LogDensityProblemsAD.jl/blob/e3401f21b5a065df0d5de38b37fad0e6650618f3/ext/LogDensityProblemsADADTypesExt.jl#L43-L50

See discussion in https://github.com/tpapp/LogDensityProblemsAD.jl/pull/29.

but tldr we can't really shortcircuit LogDensityProblemsAD.

**Comment by yebai** [comment]
@penelopeysm can you take a look at this so we can finish off #2411

**Comment by sunxd3** [comment]
🎉
